sky/backends/backend_utils.py

```diff
@@ -1228,34 +1228,67 @@ def _get_tpu_vm_pod_ips(ray_config: Dict[str, Any],
-    if rcode != 0:
-        failure_massage = ('Failed to run gcloud to get TPU VM Pod IPs.\n'
+    returncode, stdout, stderr = log_lib.run_with_log(query_cmd,
+                                                      '/dev/null',
+                                                      shell=True,
+                                                      stream_logs=False,
+                                                      require_outputs=True)
+    if returncode != 0:
+        failure_massage = ('Failed to run gcloud to get TPU VM IDs.\n'
                            '**** STDOUT ****\n'
-                           '{stdout}\n'
+                           f'{stdout}\n'
                            '**** STDERR ****\n'
-                           '{stderr}\n'
+                           f'{stderr}\n'
                            '**** CMD ****\n'
-                           '{tpuvm_cmd}')
+                           f'{query_cmd}\n')
         with ux_utils.print_exception_no_traceback():
-            raise RuntimeError(
-                failure_massage.format(stdout=stdout,
-                                       stderr=stderr,
-                                       tpuvm_cmd=tpuvm_cmd))
-    all_ips = re.findall(IP_ADDR_REGEX, stdout)
+            raise RuntimeError(failure_massage)
+    if len(stdout) == 0:
+        logger.warning('No TPU VMs found with cluster name '
+                       f'{cluster_name} in zone {zone}.')
+    if len(stdout.splitlines()) > 1:
+        logger.warning('Found more than one TPU VM with cluster name '
+                       f'{cluster_name} in zone {zone}.')
+
+    all_ips = []
+    for tpu_id in stdout.splitlines():
+        tpuvm_cmd = (f'gcloud compute tpus tpu-vm describe {tpu_id}'
+                     f' --zone {zone} --format=json')
+        returncode, stdout, stderr = log_lib.run_with_log(tpuvm_cmd,
+                                                          '/dev/null',
+                                                          shell=True,
+                                                          stream_logs=False,
+                                                          require_outputs=True)
+        if returncode != 0:
+            failure_massage = ('Failed to run gcloud tpu-vm describe.\n'
+                               '**** STDOUT ****\n'
+                               f'{stdout}\n'
+                               '**** STDERR ****\n'
+                               f'{stderr}\n'
+                               '**** CMD ****\n'
+                               f'{tpuvm_cmd}\n')
+            with ux_utils.print_exception_no_traceback():
+                raise RuntimeError(failure_massage)
+
+        tpuvm_json = json.loads(stdout)
+        if tpuvm_json['state'] != 'READY':
+            # May be a leaked preempted resource.
+            logger.warning(f'TPU VM {tpu_id} is not in READY state. '
+                           'Could be a garbage resource. Skipping...')
+            continue
```

Comment:
Will this state be different for differnet tpu_id?

Reply:
Each TPU VM or TPU Pod only maps to a single `tpu_id`. So yes, different `tpu_id` can have different states.
But when this multiple `tpu_id` situation happens, it means there is a leaked resource with the same cluster name as the current one. That's why I print the garbage resource message.

Normally there should be only one `tpu_id` returned with the below query command.
```
query_cmd = (f'gcloud compute tpus tpu-vm list --filter='
                 f'\\(labels.ray-cluster-name={cluster_name}\\) '
                 f'--zone={zone} --format=value\\(name\\)')
```